tests/e2e/graphics/graphics-test-cases.ts

```
@@ -132,6 +140,12 @@ function registerTestCases(testCases: TestCase[], screenshoter: Screenshoter, ou
 
 			// run in parallel to increase speed
 			const goldenScreenshotPromise = screenshoter.generateScreenshot(goldenPageContent);
+
+			if (previousAttempts) {
+				// If a test has previously failed then attempt to run the tests in series (one at a time).
+				await goldenScreenshotPromise;
+			}
```

Pretty good idea!

---

Yes, but should we do it inside a `try` `catch` block as we do it below?

```TS
const errors: string[] = [];
const failedPages: string[] = [];

let goldenScreenshot: PNG | null = null;

// run in parallel to increase speed
const goldenScreenshotPromise = screenshoter.generateScreenshot(goldenPageContent);

if (previousAttempts > 0) {
  // If a test has previously failed then attempt to run the tests in series (one at a time).
  try {
    goldenScreenshot = await goldenScreenshotPromise;
    writeTestDataItem('1.golden.png', PNG.sync.write(goldenScreenshot));
  } catch (e: unknown) {
    errors.push(`=== Golden page ===\n${(e as Error).message}`);
    failedPages.push('golden');
  }
}

const testScreenshotPromise = screenshoter.generateScreenshot(testPageContent);
...
if (previousAttempts === 0) {
  try {
    goldenScreenshot = await goldenScreenshotPromise;
    writeTestDataItem('1.golden.png', PNG.sync.write(goldenScreenshot));
  } catch (e: unknown) {
    errors.push(`=== Golden page ===\n${(e as Error).message}`);
    failedPages.push('golden');
  }
}
```

---

Thanks. I've now wrapped the await within a try/catch block. This ensures that the error message is outputted to the txt files with the test artefacts.

---

We also need to check for `if (previousAttempts === 0) {` below, otherwise

```ts
errors.push(`=== Golden page ===\n${(e as Error).message}`);
failedPages.push('golden');
```

can be called twice

---

I can change the first try/catch (within `previousAttempts > 0` branch) to have an empty catch statement. The second try/catch will catch this error again (promise is in rejected state) and handle the error reporting (to failedPages, and errors arrays). This saves repeating some of the code.

```typescript
if (previousAttempts) {
  try {
    // If a test has previously failed then attempt to run the tests in series (one at a time).
    await goldenScreenshotPromise;
  } catch {
    // Will be caught again later...
  }
}
```

Is that alright?

---

Yes, seems good